### DISCUSSION TITLE
Introduce latest tag: How to upgrade the upgrader?

### DISCUSSION THREAD

**Discussion by jhoelzel** [rancher/system-upgrade-controller#259] [Ideas]
How can i automatically upgrade the upgrader? 

Currently there is no easy way for me to get the current version of the upgrader, except using a shell script.  Henceforth no matter what i do, the upgrader will have to be upgraded manually.

The quick fix would be to introduce a latest tag, which in turn would enable other software, like argocd, to simply apply the latest version of it on it's introduction.

What do you think and how do you upgrade the upgrader?

**Comment by brandond** [comment]
Use of the latest tag is somewhat of an anti-pattern, but we could evaluate publishing it.

A better option might be to use something like keel to update the tag automatically:
https://keel.sh/

**Reply by jhoelzel** [reply]
for container images yes. but not for yaml!

There still should be the latest container version in the "latest yaml".

For instance [Argo](https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml) does it this way.

They use stable instead of latest, but basically i want to outsource the updates to argo so it always pulls and applies the latest version. If it fails i will get an alert anyway and otherwise it will create a heandsoff environment =)

/e:
Also keel has no support for argo apps nor propagating updates to git

**Reply by brandond** [reply]
I'm a little confused, I though you were asking for us to use the mutable `latest` tag on the Docker image. What do you mean about a "latest yaml"?

**Reply by jhoelzel** [reply]
that would in fact be an antipattern =)
My clusters are CIS hardened, so its not what i am thinking about.

What i mean is that i install the upgrader using argo, or even simple kustomize using a link like this:

https://github.com/rancher/system-upgrade-controller/releases/download/v0.13.1/system-upgrade-controller.yaml

inevitably i will have to adapt my argo manifest in order to update the version of the controller in my cluster by increasing the version of my yaml. If however there would be a "latest release" or "latest yaml" using the current image, my auto upgrader would be auto upgrading too.

As an example, lets take argocd itself. its actually not a "state laden" application but simply can, just like the upgrader, get all its state directly from the cluster. This makes it a perfect candidate for self updating. 

The way argo works is that it will periodically check if the above script has changed on github anyway, so therefore, if a yaml would be provided that resembles the latest or stable state, i have one less thing to worry about.

PS: im aware that on mayor changes my update might fail, but that is what the cluster alerts are for =)

**Reply by brandond** [reply]
You know that GH has that built in right? 
https://docs.github.com/en/repositories/releasing-projects-on-github/linking-to-releases#linking-to-the-latest-release

https://github.com/rancher/system-upgrade-controller/releases/latest/download/system-upgrade-controller.yaml

**Reply by jhoelzel** [reply]
this redirect does not work on strict systems that dont allow 301. Or simply said, curl https://github.com/rancher/system-upgrade-controller/releases/latest/download/system-upgrade-controller.yaml      is blank because of it.

However i did find a solution nonetheless, you guys are already provding the kustomization yaml in / which does exactly that
https://github.com/rancher/system-upgrade-controller/blob/master/kustomization.yaml

however it has been last updated: 2 years ago

**Reply by brandond** [reply]
We're not going to maintain our own latest tag; GH has that functionality built in and there is tooling available that will work with it. The kustomization is used as part of the release process to build the manifests that are published as artifacts and uses the tag being published; we don't regularly update it in-place.

**Reply by jhoelzel** [reply]
hm would you change your mind if i provide you with a github workflow that automatically will tag your latest release for you?

as for the kustomize file, its an interesting choice, since it seems to be the documented way to install the operator, right here in the readme: 

https://github.com/rancher/system-upgrade-controller#deploying

this could also simply be updated with a workflow.

Furthermore I am interested in the " tooling available that will work with it.", could you please point me in the right direction ?
My goal is it to have a self updating upgrade operator deployed to the cluster. Basically to have the upgrade upgrade itself.

**Reply by brandond** [reply]
No, having mutable release tags (that mirror the `latest` pattern used by docker image tags) is not a pattern that I have ever seen any other project observe, nor is it one we want to embrace.

**Comment by jhoelzel** [comment]
While i understand the stance and i will simply use the manifests provided in the master branch of this repository and let argo catch any rollout errors, here are some projects that do just that:

ArgoCD:
https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml

Percona Operator for Postgres:
https://raw.githubusercontent.com/percona/percona-postgresql-operator/main/deploy/bundle.yaml

External-DNS
https://github.com/kubernetes-sigs/external-dns/tree/master/kustomize


useable kustomization;
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - https://raw.githubusercontent.com/rancher/system-upgrade-controller/master/manifests/system-upgrade-controller.yaml

```
